bionic/descriptors/parsing.py

```diff
@@ -0,0 +1,252 @@
+the full language. However, I'm not an expert and didn't give a lot of attention to
+every library.)
+- I was sure the parsing would behave as I expected. (In particular, I was able to
+get the Lark library to parse the full planned language, but the grammar was
+ambiguous and I had trouble understanding the way Lark resolved the ambiguities.)
+- It's easy to provide clear, precise error messages when parsing fails.
+
+That said, I'm open to switching to a library if we can achieve the above goals.
+"""
+
+import re
+
+import attr
+
+from .ast import EntityNode, TupleNode
+from ..exception import MalformedDescriptorError
+from ..util import oneline
+
+
+def dnode_from_descriptor(descriptor):
+    """
+    Given a descriptor string, returns the parsed descriptor node.
+    """
+
+    return DescriptorParser().parse(descriptor)
+
+
+def entity_dnode_from_descriptor(descriptor):
+    """
+    Given an entity descriptor string, returns the parsed entity descriptor node.
+    """
+
+    dnode = dnode_from_descriptor(descriptor)
+    try:
+        dnode.to_entity_name()
+    except TypeError:
+        raise ValueError(f"Expected a valid entity name, but got {descriptor!r}")
+    return dnode
+
+
+TOKEN_PATTERN = re.compile(
+    r"(?P<whitespace>\s+)"
+    r"|(?P<name>[a-zA-Z_][a-zA-Z0-9_]*)"
+    r"|(?P<lparen>\()"
+    r"|(?P<rparen>\))"
+    r"|(?P<comma>,)"
+    r"|(?P<end>$)"
+)
+
+
+@attr.s
+class AugmentedToken:
+    """
+    A parsed token, along with some extra metadata.
+    """
+
+    token = attr.ib()
+    token_type = attr.ib()
+    start_pos = attr.ib()
+
+    def __str__(self):
+        return f"{self.token!r} (at position {self.start_pos})"
+
+
+class DescriptorParser:
+    "Parses descriptor strings into AST nodes."
+
+    def __init__(self):
+        # All initialization happens in parse().
+        pass
+
+    def parse(self, descriptor):
+        """
+        Parses a descriptor string into a descriptor node.
+
+        Not thread-safe (because it uses member fields to manage parsing state).
+        """
+
+        # Initialize parser state.
+        self._descriptor = descriptor
+        self._cur_aug_token = None
+        self._prev_aug_token = None
+        self._expr_stack = [ExprParseState()]
+
+        # Parse each token.
+        for aug_token in self._gen_aug_tokens():
+            self._cur_aug_token = aug_token
+            self._parse_cur_aug_token()
+            self._prev_aug_token = aug_token
+
+        # The current (and only) expression on the stack should now have the parsed
+        # node.
+        assert self._cur_expr.parsed_dnode is not None
+        return self._cur_expr.parsed_dnode
+
+    @property
+    def _cur_expr(self):
+        return self._expr_stack[-1]
+
+    def _gen_aug_tokens(self):
+        descriptor = self._descriptor
+        pos = 0
+        while True:
+            match = TOKEN_PATTERN.match(descriptor, pos=pos)
+            if not match:
```

Comment:
Sure, I'm down with that. In this particular case I'm more comfortable with the shorthand `if not match` because `match` is created immediately beforehand and it's pretty clear what type it has and what it means for it to be "falsy", but I'm also happy to be consistently explicit.